### PR TITLE
Adjust how we discover the "root" of the repository when booting the test-proxy

### PR DESCRIPTION
**Pull request by scbedd**
@mccoyp 

The previous method was dependent on `azure-sdk-tools` installed in `-e` mode. The fallback path ascension logic worked in _most_ cases, but on my machine it just spun forever instead of booting the docker instance.

This adjusts to use the `request` fixture -> retrieve collected tests -> ascend to repo root from one of the collected tests -> use that as the repo root.

It is much less hacky, and should work wherever we invoke tests from, even outside the repo.

### CODE REVIEW

File: tools/azure-sdk-tools/devtools_testutils/proxy_startup.py [hunk 1]
```diff
@@ -57,6 +55,30 @@ def get_image_tag() -> str:
...
+    """
+    if os.path.isfile(start_dir_or_file):
+        current_dir = os.path.dirname(start_dir_or_file)
+    else:
+        current_dir = start_dir_or_file
+
+    while current_dir is not None and not (os.path.dirname(current_dir) == current_dir):
+        possible_root = os.path.join(current_dir, ".git")
+
+        # we need to check for assets.json first!
```
Comment (scbedd) [inline tools/azure-sdk-tools/devtools_testutils/proxy_startup.py:72]:
```suggestion
```

File: tools/azure-sdk-tools/devtools_testutils/proxy_startup.py [hunk 2]
```diff
@@ -28,25 +28,23 @@
...
     version_file_location = os.path.relpath("eng/common/testproxy/target_version.txt")
-    version_file_location_from_root = os.path.abspath(os.path.join(REPO_ROOT, version_file_location))
+    version_file_location_from_root = os.path.abspath(os.path.join(repo_root, version_file_location))
 
     try:
         with open(version_file_location_from_root, "r") as f:
             image_tag = f.read().strip()
-
     # In live pipeline tests the root of the repo is in a different location relative to this file
     except FileNotFoundError:
```
Comment (scbedd) [inline tools/azure-sdk-tools/devtools_testutils/proxy_startup.py:44]: I should probably just rip this whole thing out now yeah?

File: tools/azure-sdk-tools/devtools_testutils/proxy_startup.py [hunk 3]
```diff
@@ -57,6 +55,29 @@ def get_image_tag() -> str:
...
+    while current_dir is not None and not (os.path.dirname(current_dir) == current_dir):
+        possible_root = os.path.join(current_dir, ".git")
+
+        # we need the git check to prevent ascending out of the repo
+        if os.path.exists(possible_root):
+            return current_dir
+        else:
+            current_dir = os.path.dirname(current_dir)
+
+    raise Exception('Requested target "{}" does not exist within a git repo.'.format(start_dir_or_file))
```
Comment (mccoyp) [inline tools/azure-sdk-tools/devtools_testutils/proxy_startup.py:78]: nit: I love f-strings
```suggestion
    raise Exception(f'Requested target "{start_dir_or_file}" does not exist within a git repo.')
```

File: tools/azure-sdk-tools/devtools_testutils/proxy_startup.py [hunk 4]
```diff
@@ -148,7 +171,7 @@ def start_test_proxy() -> None:
     set_custom_default_matcher(excluded_headers=headers_to_ignore)
 
 
-def stop_test_proxy() -> None:
+def stop_test_proxy(request) -> None:
```
Comment (mccoyp) [inline tools/azure-sdk-tools/devtools_testutils/proxy_startup.py:174]: Doesn't hurt to accept `request`, but it looks like we can also drop it for now

**Review by mccoyp — APPROVED** [review]

File: tools/azure-sdk-tools/devtools_testutils/proxy_startup.py (hunk 3 above)
Comment (scbedd) [reply tools/azure-sdk-tools/devtools_testutils/proxy_startup.py:78]: Know what. you're absolutely right. I should modernize my style on this.

File: tools/azure-sdk-tools/devtools_testutils/proxy_startup.py (hunk 4 above)
Comment (scbedd) [reply tools/azure-sdk-tools/devtools_testutils/proxy_startup.py:174]: Right! Cleaning it. thanks for the catch

Comment (scbedd) [inline tools/azure-sdk-tools/devtools_testutils/proxy_startup.py:174]:
```suggestion
def stop_test_proxy() -> None:
```

File: tools/azure-sdk-tools/devtools_testutils/proxy_startup.py [hunk 5]
```diff
@@ -166,13 +189,13 @@ def stop_test_proxy() -> None:
...
     if is_live_and_not_recording():
         yield
     else:
-        start_test_proxy()
+        start_test_proxy(request)
         # Everything before this yield will be run before fixtures that invoke this one are run
         # Everything after it will be run after invoking fixtures are done executing
         yield
-        stop_test_proxy()
+        stop_test_proxy(request)
```
Comment (scbedd) [inline tools/azure-sdk-tools/devtools_testutils/proxy_startup.py:201]:
```suggestion
        stop_test_proxy()
```

**Pull request merged**